pkg/controller/memberstatus/memberstatus_controller.go

```diff
@@ -0,0 +1,305 @@
+	}
+
+	reqLogger.Info(fmt.Sprintf("Finished updating MemberStatus, requeueing after %v", defaultRequeueTime))
+	return reconcile.Result{RequeueAfter: defaultRequeueTime}, nil
+}
+
+type statusHandler struct {
+	Name         statusComponentTag
+	handleStatus func(logger logr.Logger, memberStatus *toolchainv1alpha1.MemberStatus) error
+}
+
+// aggregateAndUpdateStatus runs each of the status handlers. Each status handler reports readiness for a toolchain component. If any
+// component status is not ready then it will set the condition of the top-level status of the MemberStatus resource to not ready.
+func (r *ReconcileMemberStatus) aggregateAndUpdateStatus(reqLogger logr.Logger, memberStatus *toolchainv1alpha1.MemberStatus) error {
+
+	memberOperatorStatusHandler := statusHandler{Name: memberOperator, handleStatus: r.memberOperatorHandleStatus}
+	hostConnectionStatusHandler := statusHandler{Name: hostConnection, handleStatus: r.hostConnectionHandleStatus}
+
+	statusHandlers := []statusHandler{memberOperatorStatusHandler, hostConnectionStatusHandler}
+
+	// track components that are not ready
+	unreadyComponents := []string{}
+
+	// retrieve component statuses eg. kubefed, member deployment
+	for _, handler := range statusHandlers {
+		err := handler.handleStatus(reqLogger, memberStatus)
+		if err != nil {
+			reqLogger.Error(err, "status update problem")
+			unreadyComponents = append(unreadyComponents, string(handler.Name))
+		}
+	}
+
+	// if any components were not ready then set the overall status to not ready
+	if len(unreadyComponents) > 0 {
+		err := fmt.Errorf("Components not ready: %v", unreadyComponents)
+		return r.setStatusNotReady(memberStatus, err.Error())
+	}
+	return r.setStatusReady(memberStatus)
+}
+
+// hostConnectionHandleStatus retrieves the host cluster object that represents the connection between this member cluster and the host cluster.
+// It then takes the status from the cluster object and adds it to MemberStatus. Finally, it checks its status and will return an error if
+// its status is not ready
+func (r *ReconcileMemberStatus) hostConnectionHandleStatus(reqLogger logr.Logger, memberStatus *toolchainv1alpha1.MemberStatus) error {
+	// look up host connection status
+	fedCluster, ok := r.getHostCluster()
+	if !ok {
+		notFoundMsg := errMsgHostConnectionNotFound
+		notFoundReason := reasonHostConnectionNotFound
+		memberStatus.Status.HostConnection = kubefed_v1beta1.KubeFedClusterStatus{
+			Conditions: []kubefed_v1beta1.ClusterCondition{
+				{
+					Type:          "Ready",
+					Status:        corev1.ConditionFalse,
+					Reason:        &notFoundReason,
+					Message:       &notFoundMsg,
+					LastProbeTime: metav1.Now(),
+				},
+			},
+		}
+		return fmt.Errorf(notFoundMsg)
+	}
+	memberStatus.Status.HostConnection = *fedCluster.ClusterStatus.DeepCopy()
+
+	// check conditions of host connection
+	if !util.IsClusterReady(fedCluster.ClusterStatus) {
+		return fmt.Errorf("the host connection is not ready")
+	}
```

Comment:
This is a valid point. However, the design for the toolchain status centres around the host. Its usefulness is to be able to determine the status from a single point (the toolchainstatus resource available on the host). So for the case where there's a problem with the host deployment I think it should be evident that there's a problem because its status would not have been updated for a long time?

Reply:
Member status should not care about the host operator deployment or the host operator status at all. It's for the host status to take care of it.
Member status is interested in the host kubefedcluster to make sure the host cluster is available (regardless of the host operator status in the cluster). So, yes, checking the `lastProbeTime` of the host kubefedcluster does make sense in the member status.

Reply:
yeah, I forgot that the status info will be put together in the host-operator and not added from the member-operator side. So I agree with you :+1:

Reply:
I added a check for the lastProbeTime to ensure it is within limits. It should be updated less than `(period + timeout) * failure threshold` seconds ago

Example of the lastProbeTime status condition for the hostConnection:
```
status:
  conditions:
  - lastTransitionTime: "2020-06-30T09:48:29Z"
    lastUpdatedTime: "2020-06-30T09:48:29Z"
    message: 'components not ready: [memberOperator hostConnection]'
    reason: ComponentsNotReady
    status: "False"
    type: Ready
  hostConnection:
    conditions:
    - lastProbeTime: "2020-06-30T09:48:19Z"
      lastTransitionTime: "2020-06-30T09:46:59Z"
      message: /healthz responded with ok
      reason: ClusterReady
      status: "True"
      type: Ready
    - lastProbeTime: "2020-06-30T09:48:19Z"
      message: 'exceeded the maximum duration since the last probe: 39s'
      reason: KubefedLastProbeTimeExceeded
      status: "False"
      type: ProbeWorking
    region: ""
  memberOperator:
    buildTimestamp: "2020-06-30T19:58:59Z"
    conditions:
    - lastTransitionTime: "2020-06-30T09:48:29Z"
      lastUpdatedTime: "2020-06-30T09:48:29Z"
      message: the member operator deployment has unavailable replicas
      reason: DeploymentUnavailableReplicas
      status: "False"
      type: Ready
    deployment:
      name: member-operator
    revision: e38ad2d0883e3295ae32a26d96c04dfc7a9cb57c-dirty
    version: 0.0.1
```